### PR TITLE
feat: plumb context further into diff and solver packages

### LINKED ISSUE #90: fix: pass along context based on os signal to each request
If a user sends SIGINTERM to decK, then it should be able to exit quickly.
This can be achieved by passing along context down to requests.

### PR DESCRIPTION
**Pull request by hbagdi**
This is continuation of the effort started in 7b85526.

Fix #90

### CODE REVIEW

File: crud/registry_test.go [hunk 1]
```diff
@@ -100,26 +101,26 @@ func TestRegistryCreate(t *testing.T) {
 	err := r.Register("foo", a)
 	assert.Nil(err)
 
-	res, err := r.Create("foo", "yolo")
+	res, err := r.Create(nil, "foo", "yolo")
```
Comment (mflendrich) [inline crud/registry_test.go:104]: Why `nil`?

https://golang.org/pkg/context/ says
> Do not pass a nil Context, even if a function permits it. Pass context.TODO if you are unsure about which Context to use.

Comment (hbagdi) [reply crud/registry_test.go:104]: These are unit tests, does it matter?
Maybe from a referential code perspective...

Comment (hbagdi) [reply crud/registry_test.go:104]: ping @mflendrich

**Comment by hbagdi** [comment]
Update:
The linter has now been turned on for all non-test code to catch for cases where we might be passing a nil context.

**Comment by mmorel-35** [comment]
Part 2 of #325

File: crud/registry_test.go (hunk 1 above)
Comment (mflendrich) [reply crud/registry_test.go:104]: The documentation for `context.Background` says

> It is typically used by the main function, initialization, and tests

Given that:
- the documentation says to never pass `nil`
- the documentation suggests that `context.Background` is intended for use in tests
- it's a matter of a single line `ctx := context.Background` and replacing `nil` by `ctx`
- functions that get a nil `context` will panic if they try to call any function on a nil context (and from the perspective of tests we shouldn't be assuming that they don't)

I can see strong evidence convincing that we should pass a valid context also in tests.

The cost is one `ctx := context.Background()` line per function and replacement of `nil`s with `ctx`s which doesn't look like a big deal to me.

**Review by mflendrich — APPROVED** [review]
Please see the comment

File: crud/registry_test.go (hunk 1 above)
Comment (hbagdi) [reply crud/registry_test.go]: Okay. I've updated the code. The most convincing part was `nil` context point.

**Comment by hbagdi** [comment]
I've update the PR to include changes that happen in main due to merge of Konnect documents work. Please take another look.

**Review by mflendrich — COMMENTED** [review]
please resolve the conflicts before the final LGTM because the nature of these conflicts seems particularly bug-prone to me, so a second pair of eyes could help.

**Comment by hbagdi** [comment]
@mflendrich That merge conflict took a while to resolve. PTAL.

**Comment by hbagdi** [comment]
@mflendrich Again resolved the conflicts, please review at your earliest convenience.

**Review by mflendrich — APPROVED** [review]

**Pull request merged**